### ISSUE TITLE
Allow administrators to configure rate limits

### ISSUE THREAD

**Issue by emilyst** [tootsuite/mastodon#15832]
This issue is to request the ability to change the rate limits using environment variables or using the UI (i.e., in the administration site settings).

I host a personal, single-user Mastodon instance for my own use in a Docker container. Because I'm the only one using it, registrations are closed. In other words, it is not a public instance.

I've encountered numerous situations in which I've been rate-limited from the web app. This sometimes happens if I am interacting with the web app on one computer (e.g., for work), while another one has the web app open in the background (e.g., my personal computer).

I don't necessarily want to disable rate-limiting, but I would like to make it a little more lax for my own purposes.

I don't mind throwing together a pull request for this myself, since I am familiar with Rails development, but I wanted to see how open this project is to such a change.

**Comment by weex** [comment]
It's free software so doesn't necessarily need to be merged to be helpful. I would appreciate such a PR as it opens up more configuration options, and depending on how you do it, maybe will be useful for adding other more personal mods.

**Comment by cd-sascha** [comment]
Hello, may I ask if it is planned to address this in a coming release? With the the higher popularity of mastodon we are more often running into this issue again and again, esp. if you are running behind NAT.

**Comment by emilyst** [comment]
If anyone does take this on, there is some subtle complexity to consider. Much of this is discussed in this similar issue: https://github.com/mastodon/mastodon/issues/10557.

To summarize, there are two main issues:

1. Rate limits are designed to ease federation by making it harder to DoS other instances. Without rate limits on things like posting, edits, deletions, etc., it would be easy to overwhelm federated instances with new messages and tasks.
2. Even for those rate limits which only affect local instances, there is still the risk of tying up resources with many Sidekiq jobs (which are used to do the lion's share of tasks done by Mastodon).

Therefore it would be advisable to allow adjusting rate limits _for GET requests only_, at least at first. Nothing which changes or propagates state can be easily changed without implications for performance, security, and federation.

For this reason, I am going to amend the title slightly to clarify what I want.

**Comment by phocks** [comment]
I think this might be the file that controls some rate limits. I'm wondering if I bump these numbers on my single-user instance and then re-compile, if that would do the trick. I'm basically trying to loop through my following list and make changes. Currently I'm just using an async `sleep` function to throttle requests, but because it's my own server, I feel like I should be able to raise the rate limit myself. https://github.com/mastodon/mastodon/blob/46ad7fea9d67631f54dd1ef45114a08cd2c5db73/config/initializers/rack_attack.rb

**Comment by phocks** [comment]
Can confirm, raising this number from 300 to 99999 raised the api limit for me.
https://github.com/mastodon/mastodon/blob/46ad7fea9d67631f54dd1ef45114a08cd2c5db73/config/initializers/rack_attack.rb#L49

**Comment by muratcorlu** [comment]
I started to see this problem is getting bigger with increasing popularity of mastodon servers by users behind some popular VPN-like networks like Apple’s iCloud Private Relay. iCloud Private Relay is by default enabled for all of the iCloud+ users and it can cause to use same IP address by many iPhone users.

> Traditional fraud detection that relies solely on IP addresses might need to be updated to ensure legitimate users are not impacted. Consider treating these addresses like larger carrier grade NAT or enterprise IP addresses to better account for this type of traffic, since many Private Relay users may be assigned to a single relay IP address.
> https://developer.apple.com/support/prepare-your-network-for-icloud-private-relay

I started to see false-positive rate limits on different mastodon servers that I visit for the first time. Here is an example: https://synaps.space/@murat/109455195233028760

I think, administrators should be able to at least disable IP based rate limiting. I don’t even think IP-based rate limiting in that manner helps anyone more than hurts at all.

**Comment by emilyst** [comment]
In pull request #23129, I pushed up the least possible thing that could work, as a draft pull request. I just wanted to start discussion on the implementation, and I don’t feel precious about this approach at all, so I welcome input from any maintainers.

**Comment by nfriedly** [comment]
I also am running on the Docker image to a single-user instance, and I found myself getting limited on GET requests just from having a dozen or so tabs open today. This looks like a sound improvement to me.

It does make me wonder just how chatty the app is, if it managed to hit 300 requests in 5 minutes (multiple times) just from my activity alone...

**Comment by emilyst** [comment]
Yeah, I still experience issues from this. I can’t tell what the lack of input on this is about.

**Comment by nfriedly** [comment]
> Were you able to figure it out on your docker instance?

No, I just closed most of the tabs and it started behaving again a few minutes later.

**Comment by davinkevin** [comment]
+1, one instance for one user… and I have this limit every-time I want to follow a bunch of accounts and attach them to a list… manually (not through script).

**Comment by deutrino** [comment]
This also makes it difficult to run a multimedia heavy curated feed by boosting out of favorites or bookmarks, because once you scroll far enough back, everything stops loading and you have to wait for the next wall clock quarter hour to elapse.

See also #22217, #26076, #27282, #29332, and probably others.

**Comment by nfriedly** [comment]
I feel like making multiple tabs coordinate with each other (e.g. via [BroadcastChannel](https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API)), and/or slowing the rate of refreshing [when in the background](https://developer.mozilla.org/en-US/docs/Web/API/Window/blur_event) would significantly help here.

But, yeah, ultimately the rate-limiting needs to be configurable.

**Comment by thibaultmol** [comment]
I think having configurable rate limits even per user could be interesting. For example having admins/mods get higher rate limits as they might need to open many tabs or go through many profiles/media at once

**Comment by sean666888** [comment]
I'm using FediFetcher backfill stuff and I run into rate limits all the time especially when I am actually trying to use my instance at the same time so a per user or even a per token ability to change the default API rate limits would be excellent

**Comment by deutrino** [comment]
See #16582, #23529, and PR #30447...

**Comment by flancian** [comment]
Hi there! I'm one of the admins of a medium sized instance (~1.5k monthly active users) and we get reports by people surprised at being throttled during what nowadays amounts to normal web browsing patterns pretty often. I myself hit these rate limits often while using social.coop "normally".

Could we get this prioritized or at least triaged by the Mastodon project? Would a simple PR just bumping the current presumably old school limit slightly have a chance of being accepted?

Thank you in any case!

**Comment by deutrino** [comment]
This should arguably have the "partially a bug" label if not the full "bug" label, as it not being addressed leads to broken behavior when scrolling back in Bookmarks and Favorites, and when viewing the Media tab on remote accounts.

**Comment by deutrino** [comment]
If adding a configurable option is too much work, perhaps someone could address #30447, which simply raises the arbitrarily low limit for logged-in users.
